sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/sycl_ext_oneapi_bindless_images.asciidoc

```diff
@@ -0,0 +1,1826 @@
+  raw_handle_type raw_handle;
+}
+
+class image_mem {
+
+  image_mem();
+  image_mem(const image_mem &) = delete; // no copy-construct
+
+  image_mem(image_mem &&rhs) noexcept; // move-constructor is allowed
+  image_mem(const sycl::context &syclContext, const image_descriptor &desc);
+
+  ~image_mem();
+
+  image_mem &operator=(image_mem &&); // move-assignment is allowed
+  image_mem &operator=(image_mem &) = delete; // no copy-assignment
+
+  image_mem_handle get_handle() const;
+  image_descriptor get_descriptor() const;
+  sycl::context get_context() const;
+
+  sycl::range<3> get_range() const;
+  sycl::image_channel_type get_image_channel_type() const;
+  sycl::image_channel_type get_image_channel_order() const;
+  unsigned int get_image_num_channels() const;
+  image_type get_type() const;
+
+  image_mem_handle get_mip_level() const;
+};
+
+image_mem_handle alloc_image_mem(const context& syclContext, 
+                                 const image_descriptor &desc);
+void free_image_mem(const context& syclContext, 
+                    image_mem_handle memHandle
+                    image_type imgType);
+}
+```
+
+The first method of allocating device memory for images is through 
+`alloc_image_mem`. This takes an `image_descriptor` and `sycl::context` to 
+allocate device memory, appropriately sized based on the `image_descriptor`.
+
+Memory allocated in this way requires the user to free that memory after all 
+operations using the memory are completed and no more operations operating on 
+the memory will be scheduled. This is done using `free_image_mem`. An 
+`image_type` should be passed to `free_image_mem` to inform the implementation 
+of the type of memory to be freed.
+
+The `image_mem` class is a RAII class wrapper that performs allocation and 
+deallocation of device memory.
+
+The default constructor does not allocate any memory on the device and the 
+resulting `image_mem` object is in an uninitialized state.
+
+The constructor `image_mem(const sycl::context &, const image_descriptor &)` 
+is a wrapper for `alloc_image_mem` functionality. 
+
+The destructor is a wrapper for `free_image_mem` functionality. 
+
+`image_mem` also provides some functions to get various properties of the 
+image memory allocation such as the range, channel type, channel order, number 
+of channels, number of levels, and image type.
+
+In the case where a mipmap has been allocated, `get_mip_level` can be used to 
+return an `image_mem_handle` to a specific level of the mipmap. This can then 
+be used to copy data to that specific level or create an image handle based on 
+that level.
+
+Note that the handle type `image_mem_handle::raw_handle_type` is an opaque type, 
+and the handle cannot be dereferenced on the host. The layout of the memory is 
+backend-specific, and may be an optimized layout, e.g. tile swizzle patterns.
+
+`image_mem` shall not be copy-constructible, copy-assignable, or 
+device-copyable.
```

Comment:
Many other SYCL classes have "common reference semantics".  Why not define `image_mem` that way too?  Defining it in a different way (e.g. no copy constructor) will lead to confusion.

Objects with common reference semantics have shallow copy semantics.  Therefore, copying an `image_mem` would not create a new underlying memory buffer.  Rather, you just get a new object wrapping the same buffer.

Reply:
Thanks for the clarification, I wasn't aware that common reference semantics have shallow copy semantics. We will look more into the common reference semantics and update this in the next revision.